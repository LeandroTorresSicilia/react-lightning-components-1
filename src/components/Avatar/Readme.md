Avatar examples:

    <div>
        <div className="slds-p-bottom--medium">
            <Avatar src="images/avatar1.jpg" initials="JL" size="large" title="Jose Leandro" alt="Jose Leandro" />
        </div>
        <div className="slds-p-bottom--medium">
            <Avatar src="src/path/wrong.jpg" initials="JL" title="Jose Leandro" />
        </div>
        <div className="slds-p-bottom--medium">
            <Avatar title="Jose Leandro" initials="KLl" style={{ backgroundColor: 'gray' }} variant={ true } />
        </div>
        <div className="slds-p-bottom--medium">
            <Avatar iconName="custom:custom1" title="Heart" size="small" />
        </div>
        <div className="slds-p-bottom--medium">
            <Avatar src="images/avatar2.jpg" title="Woman avatar" variant={ true } />
        </div>
    </div>