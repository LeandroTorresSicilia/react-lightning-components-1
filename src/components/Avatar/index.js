import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import Icon from '../Icon';

const INITIALS_REGEX = /^[A-Z][A-Z]$/;

export default class Avatar extends Component {
    constructor(props) {
        super(props);
        this.state = {
            imageFailed: false
        };
    }

    render() {
        const { 
            alt, 
            initials, 
            src,
            style, 
            title
        } = this.props;
        const { imageFailed } = this.state;
        const isInitialsValid = INITIALS_REGEX.test(initials);
       
        return (
                <span className={ this.getAvatarClass() } style={ style } >
                    {   
                        src && !imageFailed ?
                            <img alt={ alt } src={ src } onError={ () => this.setState({ imageFailed: true }) } title={ title } /> :
                        initials && isInitialsValid ?
                            <abbr className="slds-avatar__initials slds-icon-standard-user" style={ style } title={ title }>{ initials }</abbr> :
                            <Icon { ...this.props } size="medium" variant="bare" />      
                    }
                </span>
            );

    }

    getAvatarClass() {
        const { className, size, variant } = this.props; 

        return classnames('slds-avatar', {
            'slds-avatar_circle': variant,
            'slds-avatar_large': size === 'large',
            'slds-avatar_medium': size === 'medium',
            'slds-avatar_small': size === 'small',
            'slds-avatar_x-small': size === 'x-small'
        }, className);
    }
}

Avatar.propTypes = {
    /** Avatar description. It usually is a person name. Can be used only if valid src is available */
    alt: PropTypes.string,
    /** Is the avatar source path. Its value is the relative URL to the image. It take precedence over the initilas and icon */
    src: PropTypes.string,
    /** If set to true make the avatar a circle. */
    variant: PropTypes.bool,
    /** Class for custom styles */
    className: PropTypes.string,
    /** The user initials. It only can have two letters (the first name and last name first letter in upper case). It take precedence over the icon */
    initials: PropTypes.string,
    /** The icon name. It have the less precedence. It take the following format: ‘sprite name:icon name’ e.g. ‘utility:add’ */
    iconName: PropTypes.string,
    /** The icon size. */
    size: PropTypes.oneOf(['large', 'medium','small', 'x-small']),
    /** Object with the custom styles. The properties must be in camelCase naming convention (e.g. { backgroundColor: green }) */
    style: PropTypes.object,
    /** This is a description that is showed when a user hover the avatar */
    title: PropTypes.string
};

Avatar.defaultProps = {
    iconName: 'standard:user',
    size: 'medium',
    variant: false
};