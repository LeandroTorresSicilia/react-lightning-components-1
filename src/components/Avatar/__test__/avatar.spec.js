import React from 'react';
import Avatar from './../index';
import {shallow} from 'enzyme';
import renderer from 'react-test-renderer';

describe('<Avatar />', () => {
    it('should render a large image avatar', () => {
        const component = renderer.create(
            <Avatar src="images/avatar1.jpg" initials="JL" size="large" title="Jose Leandro" alt="Jose Leandro" />
        );

        expect(component).toMatchSnapshot();
    }); 

    it('should render a avatar with initials', () => {
        const component = renderer.create(
            <Avatar initials="JL" />
        );

        expect(component).toMatchSnapshot();
    });

    it('should render a avatar with a default user icon', () => {
        const component = renderer.create(
            <Avatar />
        );

        expect(component).toMatchSnapshot();
    });

    it('should render a avatar with a default user icon when the initials provided are wrong', () => {
        const component = renderer.create(
            <Avatar initials="JLl" />
        );

        expect(component).toMatchSnapshot();
    });

    it('should render a avatar with a custom icon', () => {
        const component = renderer.create(
            <Avatar iconName="custom:custom1" />
        );

        expect(component).toMatchSnapshot();
    });

    it('should render a avatar with a cirle style when variant is set to true', () => {
        const component = renderer.create(
            <Avatar src="images/avatar2.jpg" variant={ true } />
        );

        expect(component).toMatchSnapshot();
    });

    it('should fire onError event when src prop is not valid (image is not loaded)', done => {
        const onErrorMockFnCallback = jest.fn(() => done());
        const component = shallow(
            <Avatar src="wrong/path" onError={ onErrorMockFnCallback() } />
        );

        expect(onErrorMockFnCallback).toHaveBeenCalledTimes(1);
    });

});