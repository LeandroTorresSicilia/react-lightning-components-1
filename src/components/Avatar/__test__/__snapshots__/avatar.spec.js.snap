// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Avatar /> should render a avatar with a cirle style when variant is set to true 1`] = `
<span
  className="slds-avatar slds-avatar_circle slds-avatar_medium"
  style={undefined}
>
  <img
    alt={undefined}
    onError={[Function]}
    src="images/avatar2.jpg"
    title={undefined}
  />
</span>
`;

exports[`<Avatar /> should render a avatar with a custom icon 1`] = `
<span
  className="slds-avatar slds-avatar_medium"
  style={undefined}
>
  <span
    className="slds-icon_container slds-icon-custom-custom1"
    style={undefined}
    title={undefined}
  >
    <svg
      aria-hidden="true"
      className="slds-icon"
      style={undefined}
    >
      <use
        xlinkHref="/icons/custom-sprite/svg/symbols.svg#custom1"
        xmlnsXlink="http://www.w3.org/1999/xlink"
      />
    </svg>
    <span
      className="slds-assistive-text"
    />
  </span>
</span>
`;

exports[`<Avatar /> should render a avatar with a default user icon 1`] = `
<span
  className="slds-avatar slds-avatar_medium"
  style={undefined}
>
  <span
    className="slds-icon_container slds-icon-standard-user"
    style={undefined}
    title={undefined}
  >
    <svg
      aria-hidden="true"
      className="slds-icon"
      style={undefined}
    >
      <use
        xlinkHref="/icons/standard-sprite/svg/symbols.svg#user"
        xmlnsXlink="http://www.w3.org/1999/xlink"
      />
    </svg>
    <span
      className="slds-assistive-text"
    />
  </span>
</span>
`;

exports[`<Avatar /> should render a avatar with a default user icon when the initials provided are wrong 1`] = `
<span
  className="slds-avatar slds-avatar_medium"
  style={undefined}
>
  <span
    className="slds-icon_container slds-icon-standard-user"
    style={undefined}
    title={undefined}
  >
    <svg
      aria-hidden="true"
      className="slds-icon"
      style={undefined}
    >
      <use
        xlinkHref="/icons/standard-sprite/svg/symbols.svg#user"
        xmlnsXlink="http://www.w3.org/1999/xlink"
      />
    </svg>
    <span
      className="slds-assistive-text"
    />
  </span>
</span>
`;

exports[`<Avatar /> should render a avatar with initials 1`] = `
<span
  className="slds-avatar slds-avatar_medium"
  style={undefined}
>
  <abbr
    className="slds-avatar__initials slds-icon-standard-user"
    style={undefined}
    title={undefined}
  >
    JL
  </abbr>
</span>
`;

exports[`<Avatar /> should render a large image avatar 1`] = `
<span
  className="slds-avatar slds-avatar_large"
  style={undefined}
>
  <img
    alt="Jose Leandro"
    onError={[Function]}
    src="images/avatar1.jpg"
    title="Jose Leandro"
  />
</span>
`;
